pub/spec/dependabot/pub/update_checker_spec.rb

```diff
@@ -465,6 +468,81 @@
     end
   end
 
+  describe "#lowest_resolvable_security_fix_version" do
+    subject(:lowest_resolvable_security_fix_version) { checker.lowest_resolvable_security_fix_version }
+    let(:dependency_name) { "retry" }
+    let(:security_advisories) do
+      [
+        Dependabot::SecurityAdvisory.new(
+          dependency_name: dependency_name,
+          package_manager: "pub",
+          vulnerable_versions: ["<3.0.0"]
+        )
+      ]
+    end
+
+    context "when a newer non-vulnerable version is available" do
+      # TODO: Implement https://github.com/dependabot/dependabot-core/issues/5391, then flip "highest" to "lowest"
+      it "updates to the highest non-vulnerable version" do
+        is_expected.to eq(Gem::Version.new("3.1.0"))
+      end
+    end
+
+    # TODO: should it update indirect deps for security vulnerabilities? I assume Pub has these?
```

Comment:
Good question! Yes, it does but I'm not sure if:
- we parse the transitive dependencies
- there is a capability to update them
- there is a capability to update them if they are locked by a parent dependency

I think it's worth confirming and then documenting if we don't support them yet.

Reply:
Here's a manual run of a public repo with a transitive dep:

```
[dependabot-core-dev] ~/dependabot-core $ SECURITY_ADVISORIES='[{"dependency-name":"analyzer","patched-versions":[],"unaffected-versions":[],"affected-versions":["<= 6.0.0"]}]' bin/dry-run.rb pub dart-lang/pub-dev --dir "/app" --cache=files --dep="analyzer" # transitive dep
=> reading cloned repo from /home/dependabot/dependabot-core/tmp/dart-lang/pub-dev
=> parsing dependency files
I, [2022-11-29T17:45:28.571710 #9427]  INFO -- : Failed to infer the flutter version. Attempting to use latest stable release.
I, [2022-11-29T17:45:28.571838 #9427]  INFO -- : Checking out Flutter version stable
I, [2022-11-29T17:45:29.774632 #9427]  INFO -- : Running `flutter doctor` to install artifacts and create flutter/version.
I, [2022-11-29T17:45:32.909847 #9427]  INFO -- : Running `flutter --version`
I, [2022-11-29T17:45:33.864389 #9427]  INFO -- : Installed the Flutter SDK version: 3.3.9 with Dart 2.18.5.
=> updating 1 dependencies: analyzer

=== analyzer (4.7.0) (vulnerable 🚨)
 => checking for updates 1/1
I, [2022-11-29T17:45:34.300923 #9427]  INFO -- : Failed to infer the flutter version. Attempting to use latest stable release.
I, [2022-11-29T17:45:34.301044 #9427]  INFO -- : Checking out Flutter version stable
I, [2022-11-29T17:45:35.498491 #9427]  INFO -- : Running `flutter doctor` to install artifacts and create flutter/version.
I, [2022-11-29T17:45:38.642229 #9427]  INFO -- : Running `flutter --version`
I, [2022-11-29T17:45:39.581409 #9427]  INFO -- : Installed the Flutter SDK version: 3.3.9 with Dart 2.18.5.
 => latest available version is 5.2.0
 => there is no available non-vulnerable version
 => latest allowed version is 4.7.0
 => requirements to unlock: update_not_possible
 => requirements update strategy:
    (no security update possible 🙅‍♀️)
🌍 Total requests made: '0'
```

As expected, it throws `update_not_possible`.

Here's a similar run against that repo, but for a direct dep:
```
[dependabot-core-dev] ~/dependabot-core $ SECURITY_ADVISORIES='[{"dependency-name":"buffer","patched-versions":[],"unaffected-versions":[],"affected-versions":["<= 2.0.0"]}]' bin/dry-run.rb pub dart-lang/pub-dev --dir "/app" --cache=files --dep="buffer" # direct dep
=> reading cloned repo from /home/dependabot/dependabot-core/tmp/dart-lang/pub-dev
=> parsing dependency files
I, [2022-11-29T17:44:05.782429 #8784]  INFO -- : Failed to infer the flutter version. Attempting to use latest stable release.
I, [2022-11-29T17:44:05.782594 #8784]  INFO -- : Checking out Flutter version stable
I, [2022-11-29T17:44:06.918204 #8784]  INFO -- : Running `flutter doctor` to install artifacts and create flutter/version.
I, [2022-11-29T17:44:12.748310 #8784]  INFO -- : Running `flutter --version`
I, [2022-11-29T17:44:13.784197 #8784]  INFO -- : Installed the Flutter SDK version: 3.3.9 with Dart 2.18.5.
=> updating 1 dependencies: buffer

=== buffer (1.1.1) (vulnerable 🚨)
 => checking for updates 1/1
I, [2022-11-29T17:44:14.243087 #8784]  INFO -- : Failed to infer the flutter version. Attempting to use latest stable release.
I, [2022-11-29T17:44:14.243212 #8784]  INFO -- : Checking out Flutter version stable
I, [2022-11-29T17:44:15.484309 #8784]  INFO -- : Running `flutter doctor` to install artifacts and create flutter/version.
I, [2022-11-29T17:44:18.590065 #8784]  INFO -- : Running `flutter --version`
I, [2022-11-29T17:44:19.892184 #8784]  INFO -- : Installed the Flutter SDK version: 3.3.9 with Dart 2.18.5.
 => latest available version is 1.1.1
    (no update needed as it's already up-to-date)
🌍 Total requests made: '0'
```

I was a little surprised that ☝️ says "no update needed as already up-to-date" rather than "no update possible"... but turns out that's because of the `dry-run` script logging not first checking for vulnerable:
https://github.com/dependabot/dependabot-core/blob/c2b0cf8cd7d58f01565128ac68c72c4818d7edf3/bin/dry-run.rb#L682-L694

I'll open a separate PR to tweak that shortly.

Reply:
> I was a little surprised that ☝️ says "no update needed as already up-to-date" rather than "no non-vulnerable version available"... but turns out that's because of the dry-run script logging not first checking for vulnerable... I'll open a separate PR to tweak that shortly.

PR here: https://github.com/dependabot/dependabot-core/pull/6258